services/api-server/src/simcore_service_api_server/api/dependencies/rabbitmq.py

```diff
@@ -23,64 +22,107 @@ def get_rabbitmq_client(app: Annotated[FastAPI, Depends(get_app)]) -> RabbitMQCl
         item = JobLog(
             job_id=got.project_id,
             node_id=got.node_id,
             log_level=got.log_level,
             messages=got.messages,
         )
-        await self._queue.put(item)
+        if (callback := self._log_streamers.get(item.job_id)) is None:
+            raise HTTPException(
+                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
+                detail=f"Could not forward log because a logstreamer associated with job_id={item.job_id} was not registered",
+            )
+        await callback(item)
         return True
 
-    async def _project_done(self) -> bool:
-        task = await self._director2_api.get_computation(
-            self._project_id, self._user_id
+    async def register(
+        self, job_id: JobID, callback: Callable[[JobLog], Awaitable[None]]
+    ):
+        if job_id in self._log_streamers:
+            raise HTTPException(
+                status_code=status.HTTP_409_CONFLICT,
+                detail=f"A stream was already connected to {job_id=}. Only a single stream can be connected at the time",
+            )
+        self._log_streamers[job_id] = callback
+        await self._rabbit_client.add_topics(
+            LoggerRabbitMessage.get_channel_name(), topics=[f"{job_id}.*"]
+        )
+
+    async def deregister(self, job_id: JobID):
+        if not job_id in self._log_streamers:
+            raise HTTPException(
+                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
+                detail=f"No stream was connected to {job_id=}.",
+            )
+        await self._rabbit_client.remove_topics(
+            LoggerRabbitMessage.get_channel_name(), topics=[f"{job_id}.*"]
         )
+        del self._log_streamers[job_id]
+
+
+class LogStreamer:
+    _queue: Queue[JobLog]
+    _queue_name: str
+    _job_id: JobID
+    _user_id: PositiveInt
+    _director2_api: DirectorV2Api
+
+    def __init__(
+        self,
+        user_id: Annotated[PositiveInt, Depends(get_current_user_id)],
+        director2_api: Annotated[DirectorV2Api, Depends(get_api_client(DirectorV2Api))],
+    ):
+        self._user_id = user_id
+        self._director2_api = director2_api
+        self._queue: Queue[JobLog] = Queue()
+
+    async def register(self, job_id: JobID, log_distributor: LogDistributor):
+        self._job_id = job_id
+        await log_distributor.register(job_id, self._queue.put)
+
+    async def deregister(self, log_distributor: LogDistributor):
+        await log_distributor.deregister(self._job_id)
+
+    async def _project_done(self) -> bool:
+        task = await self._director2_api.get_computation(self._job_id, self._user_id)
         return not task.stopped is None
 
     async def log_generator(self) -> AsyncIterable[str]:
         while True:
             while self._queue.empty():
                 if await self._project_done():
                     return
-                await asyncio.sleep(5)
+                await asyncio.sleep(10)
```

Comment:
I would say it is not nessecary for now. But perhaps it could be done in the future. I think the particular amount here is not critical. Let me know if you disagree.

Reply:
OK, I will setup a constant with an informative name for now. 😄